tests/linux/test_platform_linux.cpp

```diff
@@ -554,4 +557,169 @@ INSTANTIATE_TEST_SUITE_P(
+     {"VERSION_CODENAME=hirsute"},
+     {"UBUNTU_CODENAME=hirsute"}},
+    {"Ubuntu", "21.04"}};
+
+OSReleaseTestParam parse_os_release_ubuntu2104lts_rotation = {
+    {{"VERSION=\"21.04 (Hirsute Hippo)\""},
+     {"ID=ubuntu"},
+     {"ID_LIKE=debian"},
+     {"VERSION_ID=\"21.04\""},
+     {"PRETTY_NAME=\"Ubuntu 21.04\""},
+     {"HOME_URL=\"https://www.ubuntu.com/\""},
+     {"SUPPORT_URL=\"https://help.ubuntu.com/\""},
+     {"BUG_REPORT_URL=\"https://bugs.launchpad.net/ubuntu/\""},
+     {"PRIVACY_POLICY_URL=\"https://www.ubuntu.com/legal/terms-and-policies/privacy-policy\""},
+     {"VERSION_CODENAME=hirsute"},
+     {"NAME=\"Ubuntu\""},
+     {"UBUNTU_CODENAME=hirsute"}},
+    {"Ubuntu", "21.04"}};
+
+TEST_P(OSReleaseTest, test_parse_os_release)
+{
+    auto expected = std::get<1>(GetParam());
+    const QStringList input = std::get<0>(GetParam());
+
+    auto output = multipass::platform::detail::parse_os_release(input);
+
+    EXPECT_EQ(expected.first, output.first.toStdString());
+    EXPECT_EQ(expected.second, output.second.toStdString());
+}
+
+INSTANTIATE_TEST_SUITE_P(PlatformLinux, OSReleaseTest,
+                         Values(OSReleaseTestParam{{}, {"unknown", "unknown"}}, parse_os_release_empty,
+                                parse_os_release_single_char_fields, parse_os_release_ubuntu2104lts,
+                                parse_os_release_ubuntu2104lts_rotation));
+
+TEST_F(PlatformLinux, find_os_release_none_found)
+{
+    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();
+    EXPECT_CALL(*mock_file_ops, open(_, _)).Times(2);
+
+    auto output = multipass::platform::detail::find_os_release();
+    EXPECT_EQ(output->fileName(), "");
+}
+
+TEST_F(PlatformLinux, find_os_release_dummy_file)
+{
+    const QString test_file_path = mpt::test_data_path() + "os-release_sample";
+    auto os_release_dummy = std::make_unique<QFile>(test_file_path);
+// FAIL LINT, STOP BUILD.
+    //    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();
+    //    EXPECT_CALL(*mock_file_ops, open(os_release_dummy, _)).Times(1);
+
+    auto output = multipass::platform::detail::find_os_release();
+    EXPECT_EQ(output->fileName(), test_file_path);
+}
+
+TEST_F(PlatformLinux, read_os_release_from_file_not_found)
+{
+    const std::string expected = "unknown-unknown";
+
+    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();
+    EXPECT_CALL(*mock_file_ops, open(_, _)).Times(2);
+
+    auto output = multipass::platform::detail::read_os_release();
+
+    EXPECT_EQ(expected, output);
+}
+
+TEST_F(PlatformLinux, read_os_release_from_file)
+{
+    const std::string expected = "distribution_name-distribution_release";
+    const QString test_file_path = mpt::test_data_path() + "os-release_sample";
+    auto os_release_dummy = std::make_unique<QFile>(test_file_path);
+
+    //    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();
+    //    EXPECT_CALL(*mock_file_ops, open(os_release_dummy, _)).Times(2);
```

Comment:
I have not tested the following but it seems like it should work:
```c++
std::string multipass::platform::detail::read_os_release()
{
    std::unique_ptr<QFile> os_release = find_os_release();

    QStringList os_info;
    if (os_release->isOpen())
    {
        const int buffer_size = 32;
        char buffer[buffer_size];
        QTextStream input;
        while (MP_FILEOPS.read(*os_release, buffer, sizeof(char) * buffer_size) != -1)
        {
            input << buffer;
        }

        QString line = input.readLine();
        while (!line.isNull())
        {
            os_info.append(line);
            line = input.readLine();
        }
        os_release->close();

        auto parsed = parse_os_release(os_info);

        return fmt::format("{}-{}", parsed.first, parsed.second);
    }

    return "unknown-unknown";
}
```

I have added the initial setup for this test in the suite. The above code has not been added. I feel like this routine is simple enough to be verified in code review.

Reply:
Hey @surahman, don't worry, this is very close to what we need now and I can pick it up and add commits on top to sort out the remaining tests (authorship of your commits isn't affected). I am bound to fail anticipating new difficulties unless I get my hands on it. But tests are important to check future code as much as today's code. And this is so close, it would be a pity to drop :slightly_smiling_face: 

These google test/mock specifics may be beyond your interests but, if you want to know, I can think of two possible approaches:
1. Mock all the way, including `isOpen` and `readLine` (the latter to avoid the intermediate text stream in the implementation)
2. Go back to using a temporary file after all, and use [`SetArgReferee`](https://github.com/google/googletest/blob/master/docs/reference/actions.md#side-effects) in `open` to switch the arg to your earlier sample file, and combine with the return via [DoAll](https://github.com/google/googletest/blob/master/docs/reference/actions.md#composite-actions)

Reply:
I agree with you on testing 100% of code, amongst other things it protects against unexpected behavior from changes in libraries and other dependant code. Mock testing has been the more interesting part of this exercise, but if this is a feature that needs to go out then please take over. If time is not an issue I will figure it out 😉.

Reply:
@ricab Found some idle time and very quickly built up the mock hooks for `isOpen` and `readLine`. `isOpen` is working as expected but `readLine` only recovers a single line of input. I am unsure if the `readLine` issue is related to the way I have written the mock routine or the way the test fixture is set up. Thoughts?

Reply:
Of course @surahman, please have at it! :smiley:

Reply:
@ricab something appears to be broken in this test, it does not seem to want to try and open the file:
```c++
TEST_F(PlatformLinux, find_os_release_usr_lib)
{
    const auto expected_filename = QStringLiteral("/var/lib/snapd/hostfs/usr/lib/os-release");

    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();

    InSequence seq;
    EXPECT_CALL(*mock_file_ops,
                open(Property(&QFile::fileName, Eq(expected_filename)), QIODevice::ReadOnly | QIODevice::Text))
        .WillOnce(Return(false))
        .WillOnce(Return(true));
    EXPECT_CALL(*mock_file_ops, open).Times(0); // no other open attempts

    auto output = multipass::platform::detail::find_os_release();
    EXPECT_EQ(output->fileName(), expected_filename);
}
```

```
tests/linux/test_platform_linux.cpp:674: tried expectation #0: EXPECT_CALL(*mock_file_ops, open(Property(&QFile::fileName, Eq(expected_filename)), QIODevice::ReadOnly | QIODevice::Text))...(of type QString)
         Expected: to be called once
           Actual: never called - unsatisfied and active
/home/saad/Multipass/tests/linux/test_platform_linux.cpp:678: tried expectation #1: EXPECT_CALL(*mock_file_ops, open)...
         Expected: all pre-requisites are satisfied
           Actual: the following immediate pre-requisites are not satisfied:
linux/test_platform_linux.cpp:674: pre-requisite #0
                   (end of pre-requisites)
         Expected: to be never called
           Actual: never called - saturated and active
```
Any ideas?

Reply:
All builds and tests passing locally. I have pushed for review.